### ISSUE TITLE
Push Notifications not working in SDK 8.1 - React Native

### ISSUE THREAD

**Issue by jsmac13** [forcedotcom/SalesforceMobileSDK-iOS#3225]
Please fill out the following details:

1. Version of Mobile SDK Used: 8.1
2. Issue found in Native App or Hybrid App: React Native - used forcereact for project creation.
3. OS Version: all
4. Device: iPhone
5. Steps to reproduce:  create a new project with forcereact - uncomment the push notification code in AppDelegate.m
6. Actual behavior: The device does not prompt to enable push notifications.  Does register with connected app properly.  Test push notification from connected app sends successfully.
7. Expected Behavior: The device should prompt user to accept push notifications.  Push notifications should work.
8. Error Log: No errors!

This was initially discovered in a custom application build with a previous sdk and upgraded.  However, with no customization beyond changing out the ClientId, login URL, etc. to enable the notifications, the notifications do not work.

**Comment by bbirman** [comment]
@jsmac13 thanks for reporting. Please see this PR for code you can add to your app delegate to get the prompt: https://github.com/forcedotcom/SalesforceMobileSDK-Templates/pull/275

**Comment by jsmac13** [comment]
Hi @bbirman.  I put together a similar approach a couple of days ago.  This absolutely works when sending a test application from the Connected App.  However, when I try to send to the Connected App from Apex.  It gets lost.  I opened a case, but that didn't get me very far.

I have this test harness running in dev console:
Set<String> userIdSet = new Set<String>();
userIdSet.add('<ID GOES HERE>');
Map<String, Object> iOSPayload = Messaging.PushNotificationPayload.apple('test2', '', null, null);
Messaging.PushNotification msg = new Messaging.PushNotification();
msg.setPayload(iOSPayload);
msg.send('<Name of Connecteted App Goes Here>', userIdSet);

It runs successfully.  I get one of these:  Number of Mobile Apex push calls: 1 out of 10.  However, it doesn't come out on the other side.  This is where I started to get really confused and rebuilt it all from scratch.  Scratch App = Old App behavior-wise.

I truly appreciate the quick turnaround.  Any additional help here would make my year...  Full disclosure, I work for a SF partner.  This app is for a well-licensed SF Customer.  All levels of management on both sides of this equation have me directly in the cross-hairs.  Help me Obi-Wan, you're my only hope!

**Comment by bbirman** [comment]
I tried your Apex sample works on with the latest template and didn't see any issues. So far the one possibility for the Apex case is that a device isn't registered for the user. You'd still see the count log but the push wouldn't go out if if there's no device associated with the user ID

**Comment by jsmac13** [comment]
Hi @bbirman Thank you so much for testing this for me.  I realized that if it works for you, it should work for me.  The one thing I hadn't tried was setting this up in a different Salesforce Org.  I created a new developer sandbox, moved everything over, and it worked the first try.  Thank you for the code change. I do think that piece was needed.  It does work.  Now I just have to figure out what has happened to the original org.  Thanks again.  I appreciate the time and the quick turnaround!!

**Comment by bbirman** [comment]
Thanks again for letting us know about the code issue!

**Comment by jsmac13** [comment]
@bbirman Another follow-up.  The push notifications from the previously provided Apex code do work if the SFDCAuthLoginHost is either login.salesforce.com or test.salesforce.com.  However, they do not work if you set the SDK to authenticate to a Salesforce community host.  For example mytestcommunity.force.com/communityname will allow the user to login, register with the connected app, and receive a test notification from the connected app.  The Apex test will not go through.  I've tested in multiple orgs and sandboxes and the behavior seems to be consistent across all orgs.  We are pursuing a solution through other channels, but I thought I'd let you know.

**Comment by bhariharan** [comment]
@jsmac13 Thanks for providing more details. A few things to note:

- Apex push notifications are not scoped by community, they're scoped by org. What this means is that push notifications within that org would go to that user within the org, not just that community.
- Sending test notifications and Apex push notifications use different channels. Test notifications send a push to the registered device token, whereas Apex sends a push to that particular `userId`.
- A good next step would be to enable verbose logging in the developer console and look at the logs. Apex should spew out some log lines if there is no device registered for that user.
- It would be good to verify that the registration flow worked for that particular user as well.

Hope this helps.

**Comment by jsmac13** [comment]
@bhariharan I agree with all of this.  It should work this way.  The registration flow is working.  If I view mobile registrations from the user record, the newly created/updated device token is there every time.  I couldn't send a test notification from the connected app unless it were.  The Apex does not generate any errors even at the finest detail.  It doesn't give me the good messages either.  With a user logging in via test.salesforce.com, I can see the device token in the log.  From a community auth host, I cannot.  

The code I'm using for iOS is 100% Salesforce Mobile SDK 'as is'.  I changed the SDK from 7.3.3 to 8.1 because of the RN webview reference Apple no longer allows.  It 1) No longer prompts the user without the code added to the SDK and 2) is now broken for Apex push notifications.  We didn't update either side of the code as it relates to notifications.  The Apex notifications with 7.3.3 have been working since before July of 2019.

Two identical code bases.  One has the login host as test.salesforce.com, one has a community host.  The first runs as expected, the second does not.  I've verified the results in a production org, 2 sandboxes, and a development org.  There appears to be some sort of disconnect when using the community as a login host as outlined [here](https://developer.salesforce.com/docs/atlas.en-us.noversion.mobile_sdk.meta/mobile_sdk/communities_tutorial_test.htm).

There is a Salesforce Internal Case to get a specialist assigned.  If an answer is found, I'll be sure to post it here.
